### PR TITLE
Add poverty and inequality functions to MicroSeries and MicroDataFrame

### PR DESCRIPTION
**Pull request by MaxGhenis**
Still needs tests and to avoid calling legacy `microdf` functions.

### CODE REVIEW

**Comment by nikhilwoodruff** [comment]
Looks good, wonder if there's some benefit in adding in some utility code that automatically parses the inputs as either column names or arrays. Would be useful where poverty thresholds are just numbers (e.g. equivalised household income has a constant poverty threshold, this would be broadcasted over everything so df.poverty_rate("equiv...", 253) etc. would work). Could this be done with a decorator?

**Comment by MaxGhenis** [comment]
> Looks good, wonder if there's some benefit in adding in some utility code that automatically parses the inputs as either column names or arrays. Would be useful where poverty thresholds are just numbers (e.g. equivalised household income has a constant poverty threshold, this would be broadcasted over everything so df.poverty_rate("equiv...", 253) etc. would work). Could this be done with a decorator?

Yeah that'd be nice.

It'd also be good to have the poverty functions available for numpy arrays with the same feature, e.g. to work with openfisca outputs without having to create a `MicroDataFrame`,

**Comment by nikhilwoodruff** [comment]
> > Looks good, wonder if there's some benefit in adding in some utility code that automatically parses the inputs as either column names or arrays. Would be useful where poverty thresholds are just numbers (e.g. equivalised household income has a constant poverty threshold, this would be broadcasted over everything so df.poverty_rate("equiv...", 253) etc. would work). Could this be done with a decorator?
> 
> Yeah that'd be nice.
> 
> It'd also be good to have the poverty functions available for numpy arrays with the same feature, e.g. to work with openfisca outputs without having to create a `MicroDataFrame`,

Yeah. Though I still like the idea of openfisca-uk returning MicroSeries and MicroDataFrames by default.

File: microdf/generic.py [hunk 1]
```diff
@@ -207,6 +318,64 @@ def set_weight_col(self, column):
...
+
+        :param income: Column indicating income.
+        :type income: str
+        :param threshold: Column indicating threshold.
+        :type threshold: str
+        :return: Poverty rate between zero and one.
+        :rtype: float
+        """
+        pov = self[income] < self[threshold]
+        return (pov * self.weights).sum() / self.sum()
```
Comment (nikhilwoodruff) [inline microdf/generic.py:333]: Should ```self.sum()``` here be instead ```self.weights.sum()```?

**Comment by MaxGhenis** [comment]
> Yeah. Though I still like the idea of openfisca-uk returning MicroSeries and MicroDataFrames by default.

PSL antitrust violation, I love it!

**Comment by MaxGhenis** [comment]
Gotta look into why `gini` is negative

**Comment by MaxGhenis** [comment]
@nikhilwoodruff would you be up to take a look at why the `gini` is returning the negative value of what it should be? It's basically the same code as what's currently correct, but this expression is producing a negative number. Just want to check if something might be off with the slicing:

https://github.com/PSLmodels/microdf/pull/164/files#diff-6f0aaed45810071201db7256dd57f31fda3a7135a2c3bba5ad7d9f2c029b047dR123
```
np.sum(cumxw[1:] * cumw[:-1] - cumxw[:-1] * cumw[1:])
```

File: microdf/generic.py [hunk 2]
```diff
@@ -88,13 +87,125 @@ def quantile(self, quantiles):
...
+        :returns: Gini index.
+        :rtype: float
+        """
+        x = np.array(self).astype("float")
+        if negatives == "zero":
+            x[x < 0] = 0
+        if negatives == "shift" and np.amin(x) < 0:
+            x -= np.amin(x)
+        if self.weights is not None:  # Check if it's not np.ones instead?
+            sorted_indices = np.argsort(self.weights)
```
Comment (nikhilwoodruff) [inline microdf/generic.py:118]:
```suggestion
            sorted_indices = np.argsort(self)
```
Fixes the gini test for me.

**Pull request merged**